content/entry-hash/index.md

```diff
@@ -0,0 +1,219 @@
+    to _hashList_.
+12. Concatenate the elements of _hashList_ in order (i.e. `[numberHash,
+    keyHash, timestampHash, itemsHash]`), tag it with `l` (List) and hash the result.
+
+
+To walk through the algorithm I'll use the following entry:
+
+```elm
+Entry
+  { number = 6
+  , key = ID "GB"
+  , timestamp = Timestamp 2016 4 5 13 23 5 Utc
+  , item = Set [Sha256 "6b18693874513ba13da54d61aafa7cad0c8f5573f3431d6f1c04b07ddb27d6bb"]
+  }
+```
+
+And this is the current serialisation in JSON:
+
+```json
+[
+  {
+    "index-entry-number":"6",
+    "entry-number":"6",
+    "entry-timestamp":"2016-04-05T13:23:05Z",
+    "key":"GB",
+    "item-hash":["sha-256:6b18693874513ba13da54d61aafa7cad0c8f5573f3431d6f1c04b07ddb27d6bb"]
+  }
+]
+```
+
+Notice they are not exaclty the same even thought they represent the same
+entity. Also notice that the `ID` type is not yet an approved RFC. For this
+RFC this shouldn't be a problem as it works the same as if the type were a
+string.
+
+Each value MUST be tagged by a byte that identifies the type, in a similar way
+as the [objecthash](https://github.com/benlaurie/objecthash) algorithm.
+
+```elm
+type Tag
+  = Hash
+  | Integer
+  | Set
+  | String
+  | Timestamp
+  | List
+
+tagToChar : Tag -> Char
+tagToChar tag =
+  case tag of
+    Hash ->
+      'r'
+    Integer ->
+      'i'
+    Set ->
+      's'
+    String ->
+      'u'
+    Timestamp ->
+      't'
+    List ->
+      'l'
+```
+
+```elm
+hashValue : Alg -> String -> Tag -> Hash
+```
+
+Where:
+
+* `Alg` is the hashing algorithm (e.g. `Sha256`),
+* `String` is the string representation of the value to hash,
+* `Tag` is the tag for the `Value` type and
+* `Hash` is the list of bytes resulting of the hashing algorithm.
+
+Following the JSON example from above, these are the hashes for each value:
+
+```elm
+numberValue = toString 6
+numberValue == "6"
+
+numberHash = hashValue Sha256 numberValue Integer
+toHex numberHash == "396ee89382efc154e95d7875976cce373a797fe93687ca8a27589116644c4bcd"
+```
+
+```elm
+timestampValue = toString (Timestamp 2016 4 5 13 23 5 Utc)
+toHex timestampValue == "2016-04-05T13:23:05Z"
+
+timestampHash = hashValue Sha256 timestampValue Timestamp
+toHex timestampHash == "f22ecc4464f22c8fee624769189665a0afd7ef10a2775a000082c47cbd9f6419"
+```
+
+```elm
+keyValue = toString (ID "GB")
+keyValue == "GB"
+
+keyHash = hashValue Sha256 keyValue String
+toHex keyHash == "fff7021c7df4426be0f9a3c83f236eb6f85d159e624b010d65e6dde267889c21"
+```
+
+The item is a special case:
+
+It is a set of values and each value is already hashed. This means that the
+content of this set only needs a sorted concatenation. Then, tag it as a set
+`s` and hash the result.
+
+```elm
+hashValueSet : Alg -> Set Hash -> Tag -> Hash
+
+itemHash = hashValueSet Sha256 (Set [Sha256 "6b18693874513ba13da54d61aafa7cad0c8f5573f3431d6f1c04b07ddb27d6bb"]) Hash
+toHex itemHash == "94b34bd5705b1a8c31796c00347dad2f5a1364050a1f6b768c7ac289502f62ce"
```

Comment:
Can you clarify how you got this result? I have tried working through the example and can't get the same result.

Reply:
I've got the same result as you now and realise where the confusion lies.

In the step by step description you say this:

For each value item in the items set:
- Let _items_ be the set of item hashes as bytes. Note that
   the item hashes must not have the prefix (i.e. `sha-256:`) when transformed to
   bytes.
- Let _hashList_ be an empty list.
- Let _itemsBytes_ be
- For each value _item_ in the _items_ set:
- Apply the hashValue function to item **tagged with r (Hash)** and append the result to itemsBytes.
- Sort the elements of itemsBytes, concatenate them and
- Apply the hashValue function tagged with s (Set). And append the result to hashList.

From this, I did the following for the example:
1. Get the bytes for "6b18693874513ba13da54d61aafa7cad0c8f5573f3431d6f1c04b07ddb27d6bb"
2. Prepend the result of 1 with "r"
3. Hash the result of 2
4. Get the bytes of the result from 3 and prepend it with "s"
5. Hash the result of 4

And I got the result "cff910f74878650a3cceb54039bdb62707de9d20e80d4385127732a4e444bd57".

If I instead do:
1. Get the bytes for "6b18693874513ba13da54d61aafa7cad0c8f5573f3431d6f1c04b07ddb27d6bb"
2. Get the result from 1 and prepend it with "s"
3. Hash the result of 2

And I get your result of "94b34bd5705b1a8c31796c00347dad2f5a1364050a1f6b768c7ac289502f62ce".

To get your answer it appears to miss out the below step from your description:

- Apply the hashValue function to item **tagged with r (Hash)** and append the result to itemsBytes.

Does this need to be removed?

Reply:
The error was in the example, I wasn't tagging the raw values with `r`. I reviewed the original objecthash implementation and the algorithm is correct.

So, the values you were getting were right. I amended and hopefully clarified the examples a bit more.